compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs

```diff
@@ -290,12 +292,43 @@ getTypeClassDictionary world = \case
     _ ->
         Nothing
 
-simplifyExpr :: World -> Expr -> Expr
-simplifyExpr world = fst . cata go
+simplifyExpr :: Expr -> Simplifier Expr
+simplifyExpr = fmap fst . cata go'
   where
-
-    go :: ExprF (Expr, Info) -> (Expr, Info)
-    go = \case
+    go' :: ExprF (Simplifier (Expr, Info)) -> Simplifier (Expr, Info)
+    go' ms = do
+        es <- sequence ms
+        world <- gets sWorldExtended
+        let v' = freeVarsStep (fmap (freeVars . snd) es)
+        if freeVarsNull v'
+          then pure (go world es)
+          else do -- constant lifting
+            version <- gets sVersion
+            es' <- forM es $ \case
+              (e,i)
+                | freeVarsNull (freeVars i)
+                , isWorthLifting e
+                , Right ty <- runGamma world version (typeOf' e)
+                -> do
+                    name <- freshExprVarName
+                    addDefValue DefValue -- todo cacheing??
+                        { dvalBinder = (name, ty)
+                        , dvalBody = e
+                        , dvalLocation = Nothing
+                        , dvalNoPartyLiterals = HasNoPartyLiterals True -- FIXME
```

Comment:
This looks like we probably need to address it before merging? Otherwise you will produce invalid LF.

Reply:
Yes -- Will do :-)